Try waiting for db queries to finish during db shutdown

optimumtact opened the pull request:
I'm hella unsure if this will help, but I think it's worth trying.

optimumtact committed:
Try waiting for db queries to finish during db shutdown

optimumtact committed:
Fix bad define

optimumtact committed:
Better clean up the queries afterwards

MrStonedOne commented on code/controllers/subsystem/dbcore.dm:
```diff
@@ -170,17 +170,26 @@ SUBSYSTEM_DEF(dbcore)
+			qdel(query)
+		for(var/datum/db_query/query in queries_active)
+			//Finish any remaining active qeries
+			UNTIL(query.process())
```
https://github.com/tgstation/tgstation/pull/77311/files#diff-d0edfd88dc5eb1e319e6da1e72635d09dc5b35b930077ba34b67606bee32212cR600-R602 sadly there is one case where it will just always return without a true value and this will perma hang.

MrStonedOne replied:
also if this takes longer than 30 seconds for some stupid reason its probably better to just eat the crash.

optimumtact replied:
ah dick

optimumtact replied:
I've changed it to return True when the status actually is finished as well.

optimumtact replied:
i think ti's impossible for any queries to be on the queue with that state anyway, but better safe than sorry.

optimumtact committed:
Lets check our work maybe

optimumtact committed:
Refactor query process to return TRUE when the query is finished

optimumtact committed:
Rah rah rah rah rah

optimumtact committed:
We cant use execute here, these have already technically been executed

Attempting to call it again will cause them to throw an error.

Instead we just run_query_sync them until they're done

optimumtact commented on code/controllers/subsystem/dbcore.dm:
```diff
@@ -170,17 +170,26 @@ SUBSYSTEM_DEF(dbcore)
 		for(var/datum/db_query/query in queries_standby)
-			run_query(query)
+			run_query_sync(query)
+			qdel(query)
```
i'm not sure it makes sense to do this, because somethign else out there might be holding a ref to this query too.

But I also can't think of what would still be out there trying to run queries during world shutdown.

only thing i can think off is stuff happening when players connect.

optimumtact committed:
Control if queries can be made during shutdown

optimumtact committed:
Flarped up list parsing

optimumtact committed:
Dont delete queries, just remove them from the list

optimumtact committed:
PHP strings huh

optimumtact commented:
it didn't help either

san7890 commented:
is this still useful to have if it doesn't rectify the tgs crashing issue?

optimumtact commented:
it's still useful, even if it has no relevance.

The crash actually seems to be in url decoding, so maybe we need to swap chat messaging to built in url decode when rust-g is shutdown.

MrStonedOne commented:
> The crash actually seems to be in url decoding, so maybe we need to swap chat messaging to built in url decode when rust-g is shutdown.

IIRC byond made url decoding/encoding faster after we switched so it probably won't be too bad.
